### PR TITLE
Propagate Context in Pipelines

### PR DESCRIPTION
**Pull request by rahul-tuli**
Add support for using `MultiModelEngine` with `Pipeline` when `Context` is passed

Needed by #410

### CODE REVIEW

File: src/deepsparse/pipeline.py [hunk 1]
```diff
@@ -119,6 +120,7 @@ def __init__(
         scheduler: Scheduler = None,
         input_shapes: List[List[int]] = None,
         alias: Optional[str] = None,
+        context: Optional["Context"] = None,
```
Comment (mgoin) [inline src/deepsparse/pipeline.py:123]: why are the quotes needed here? looks like Context is already imported

**Review by mgoin — COMMENTED** [review]
looks good to me

File: src/deepsparse/pipeline.py [hunk 2]
```diff
@@ -301,10 +304,15 @@ def _register_pipeline_tasks_decorator(pipeline_class: Pipeline):
...
-    def from_config(cls, config: Union["PipelineConfig", str, Path]) -> "Pipeline":
+    def from_config(
+        cls,
+        config: Union["PipelineConfig", str, Path],
+        context: Optional[Context] = None,
+    ) -> "Pipeline":
         """
         :param config: PipelineConfig object, filepath to a json serialized
             PipelineConfig, or raw string of a json serialized PipelineConfig
+        :param context: Optional Context object for MultiModelEngine
```
Comment (bfineran) [inline src/deepsparse/pipeline.py:315]: same update here

**Review by bfineran — COMMENTED** [review]
looks good. few quick comments

File: src/deepsparse/pipeline.py (hunk 1 above)
Comment (rahul-tuli) [reply src/deepsparse/pipeline.py:123]: Good catch, thanks!

File: src/deepsparse/pipeline.py [hunk 3]
```diff
@@ -108,6 +108,11 @@ class PipelineImplementation(Pipeline):
         to use model as-is. Default is None
     :param alias: optional name to give this pipeline instance, useful when
         inferencing with multiple models. Default is None
+    :param context: Optional Context object to use for creating instances of
+        MultiModelEngine. The Context contains a shared scheduler along with
+        other runtime information that will be used across instances of the
+        MultiModelEngine to provide optimal performance when running multiple
+        models concurrently.
```
Comment (bfineran) [inline src/deepsparse/pipeline.py:115]: nit: remove `.` for consistency

File: src/deepsparse/pipeline.py [hunk 4]
```diff
@@ -133,6 +139,12 @@ def __init__(
         if engine_type.lower() == DEEPSPARSE_ENGINE:
             self._engine_args["scheduler"] = scheduler
 
+        self.context = context
+        if self.context and self.context.num_cores != num_cores:
```
Comment (bfineran) [inline src/deepsparse/pipeline.py:143]: if num_cores is the default (aka `None`) let's not raise an error and just override num_cores to that of the context

Comment (bfineran) [reply src/deepsparse/pipeline.py:143]: this does have the potential to miss unintended edge cases though where the user could intentionally pass in None for `num_cores` and something else for `Context.num_cores`.  Don't want to overengineer, can discuss solutions if you think this is an issue

File: src/deepsparse/pipeline.py [hunk 5]
```diff
@@ -301,10 +320,19 @@ def _register_pipeline_tasks_decorator(pipeline_class: Pipeline):
...
+        context: Optional[Context] = None,
+    ) -> "Pipeline":
         """
         :param config: PipelineConfig object, filepath to a json serialized
             PipelineConfig, or raw string of a json serialized PipelineConfig
+        :param context: Optional Context object to use for creating instances of
+            MultiModelEngine. The Context contains a shared scheduler along with
+            other runtime information that will be used across instances of the
+            MultiModelEngine to provide optimal performance when running
+            multiple models concurrently.
```
Comment (bfineran) [inline src/deepsparse/pipeline.py:335]: nit: remove . for consistency

**Review by bfineran — COMMENTED** [review]
few quick comments, updated docstring reads great

File: src/deepsparse/pipeline.py [hunk 6]
```diff
@@ -119,12 +124,22 @@ def __init__(
...
         self._model_path = model_path
         self._engine_type = engine_type
         self._alias = alias
 
+        self.context = context
+        if self.context is not None:
+            num_cores = num_cores or self.context.num_cores
+            if self.context.num_cores != num_cores:
+                raise ValueError(
+                    f"num_cores mismatch. Expected {self.context.num_cores}, "
```
Comment (bfineran) [inline src/deepsparse/pipeline.py:139]: should specify mismatch is between context and Pipeline. would be hard for a user to figure out where the expected val is coming from

**Review by bfineran — APPROVED** [review]

**Review by dbogunowicz — APPROVED** [review]

**Pull request merged**